OpenRA.Mods.Common/Traits/World/PathFinder.cs

```diff
@@ -22,66 +21,85 @@ namespace OpenRA.Mods.Common.Traits
 	[Desc("Calculates routes for mobile units based on the A* search algorithm.", " Attach this to the world actor.")]
 	public class PathFinderInfo : ITraitInfo
 	{
-		public object Create(ActorInitializer init) { return new PathFinder(init.World); }
+		public object Create(ActorInitializer init) { return new PathFinder(init.World, new NonCachedPathFinder(init.World)); }
 	}
 
-	public class PathFinder
+	public interface IPathFinder
 	{
-		const int MaxPathAge = 50;	/* x 40ms ticks */
-		static readonly List<CPos> EmptyPath = new List<CPos>(0);
+		List<CPos> FindUnitPath(CPos from, CPos target, Actor self);
+		List<CPos> FindUnitPathToRange(CPos src, SubCell srcSub, WPos target, WRange range, Actor self);
+		List<CPos> FindPath(PathSearch search);
+		List<CPos> FindBidiPath(PathSearch fromSrc, PathSearch fromDest);
+	}
 
-		readonly World world;
-		public PathFinder(World world) { this.world = world; }
+	public class PathFinder : IPathFinder
+	{
+		const int MaxPathAge = 50;	/* x 40ms ticks */
 
 		class CachedPath
 		{
-			public CPos From;
-			public CPos To;
-			public List<CPos> Result;
-			public int Tick;
-			public Actor Actor;
+			public CPos From { get; set; }
+			public CPos To { get; set; }
+			public List<CPos> Result { get; set; }
+			public int Tick { get; set; }
+			public Actor Actor { get; set; }
```

Comment:
Fields please. The only technical reason for "automatic properties always" is to avoid an ABI break when the field forms part of the assembly interface -- in the field case, we'd use the `ldfld` / `stfld` IL opcodes, whereas in the property case this is desugared to method calls. This caching mechanism is purely an implementation detail, so there is no ABI impact.

These fields should in fact all be readonly, and set only in the constructor -- a cached path does not change after it is added to the cache.